File: Algorithms/p377_Combination_Sum_IV_1.cpp
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<int> table(target+1,-1);
        return getCombinationSum4(table,nums,target);
    }
    int getCombinationSum4(vector<int>& table, vector<int>& nums, int target) {
        if(table[target]==-1){
            table[target] = 0;
            for(int i=0;i<nums.size();++i){
                if(target>nums[i]){
                    table[target] += getCombinationSum4(table,nums,target-nums[i]);
                }else if(target==nums[i]){
                    table[target] += 1;
                }
            }
        }
        return table[target];
    }
};

File: Algorithms/p377_Combination_Sum_IV/p377_Combination_Sum_IV_1.py
class Solution(object):
    def combinationSum4(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        table = [None]*(target+1)
        return self.getCombinationSum4(table,nums,target)
        
    def getCombinationSum4(self, table, nums, target):
        if table[target]==None:
            table[target] = 0
            for v in nums:
                if target>v:
                    table[target] += self.getCombinationSum4(table,nums,target-v)
                elif target==v:
                    table[target] += 1
        return table[target]
